File: Windows/Pauselet.Core.Tests/ExternalFireTests.cs
using NodaTime;
using Pauselet.Core;
using Xunit;

namespace Pauselet.Core.Tests;

/// <summary>
/// Tests for <c>RecordExternalFire</c>, the reconciliation entry point used on
/// iOS where the system delivers pre-scheduled notifications and alarms while
/// the app is not running. It must produce the state a live <c>Tick()</c>
/// would have, and it must be safe to call repeatedly with the same facts.
/// </summary>
public class ExternalFireTests
{
    private static readonly DateTimeZone Utc = DateTimeZone.Utc;

    private static Instant Date(int year, int month, int day, int hour = 0, int minute = 0) =>
        TestDates.At(Utc, year, month, day, hour, minute);

    private static (ReminderEngine, MutableDateProvider) MakeEngine(
        IReadOnlyList<Reminder> reminders, Instant now)
    {
        var store = new InMemoryDataStore(new AppData
        {
            Reminders = reminders, Settings = new Settings(), Events = [],
        });
        var clock = new MutableDateProvider(now);
        var engine = new ReminderEngine(store, clock, null, Utc);
        return (engine, clock);
    }

    [Fact]
    public void StampsAnchorAndRecordsFiredEvent()
    {
        var created = Date(2026, 3, 10, 9, 0);
        var reminder = new Reminder
        {
            Title = "Water", Schedule = new Schedule.Interval(60), CreatedAt = created,
        };
        var (engine, _) = MakeEngine([reminder], Date(2026, 3, 10, 11, 0));

        var stamp = Date(2026, 3, 10, 10, 0);
        engine.RecordExternalFire(reminder.Id, stamp);

        Assert.Equal(stamp, engine.ReminderWithId(reminder.Id)?.LastFiredAt);
        Assert.Equal(
            1, engine.Events.Count(e => e.EventOutcome == ReminderEvent.Outcome.Fired)
        );
        Assert.Equal(stamp, engine.Events.First().Date);
    }

    [Fact]
    public void RepeatedReconciliationDoesNotDuplicate()
    {
        var created = Date(2026, 3, 10, 9, 0);
        var reminder = new Reminder
        {
            Title = "Water", Schedule = new Schedule.Interval(60), CreatedAt = created,
        };
        var (engine, _) = MakeEngine([reminder], Date(2026, 3, 10, 11, 0));

        var stamp = Date(2026, 3, 10, 10, 0);
        engine.RecordExternalFire(reminder.Id, stamp);
        engine.RecordExternalFire(reminder.Id, stamp);
        engine.RecordExternalFire(reminder.Id, stamp);

        Assert.Equal(
            1, engine.Events.Count(e => e.EventOutcome == ReminderEvent.Outcome.Fired)
        );
    }

    [Fact]
    public void OlderStampThanExistingAnchorIsIgnored()
    {
        var created = Date(2026, 3, 10, 9, 0);
        var reminder = new Reminder
        {
            Title = "Water", Schedule = new Schedule.Interval(60), CreatedAt = created,
        } with { LastFiredAt = Date(2026, 3, 10, 11, 0) };
        var (engine, _) = MakeEngine([reminder], Date(2026, 3, 10, 11, 30));

        engine.RecordExternalFire(reminder.Id, Date(2026, 3, 10, 10, 0));

        // An anchor must never move backwards.
        Assert.Equal(Date(2026, 3, 10, 11, 0), engine.ReminderWithId(reminder.Id)?.LastFiredAt);
        Assert.Empty(engine.Events);
    }

    [Fact]
    public void ElapsedSnoozeIsConsumedByTheFireThatHonouredIt()
    {
        var created = Date(2026, 3, 10, 9, 0);
        var reminder = new Reminder
        {
            Title = "Tilt", Schedule = new Schedule.Interval(60), CreatedAt = created,
        } with { SnoozedUntil = Date(2026, 3, 10, 9, 40) };
        var (engine, _) = MakeEngine([reminder], Date(2026, 3, 10, 10, 0));

        engine.RecordExternalFire(reminder.Id, Date(2026, 3, 10, 9, 40));

        Assert.Null(engine.ReminderWithId(reminder.Id)?.SnoozedUntil);
        Assert.Equal(
            Date(2026, 3, 10, 9, 40), engine.ReminderWithId(reminder.Id)?.LastFiredAt
        );
    }

    [Fact]
    public void FutureSnoozeSurvivesAnEarlierExternalFire()
    {
        var created = Date(2026, 3, 10, 9, 0);
        var reminder = new Reminder
        {
            Title = "Tilt", Schedule = new Schedule.Interval(60), CreatedAt = created,
            // The user snoozed *after* the notification was delivered; the
            // snooze is a promise about the future and must not be consumed by
            // the past fire being reconciled.
        } with { SnoozedUntil = Date(2026, 3, 10, 12, 0) };
        var (engine, _) = MakeEngine([reminder], Date(2026, 3, 10, 10, 30));

        engine.RecordExternalFire(reminder.Id, Date(2026, 3, 10, 10, 0));

        Assert.Equal(
            Date(2026, 3, 10, 12, 0), engine.ReminderWithId(reminder.Id)?.SnoozedUntil
        );
    }

    [Fact]
    public void UnknownReminderIsANoOp()
    {
        var (engine, _) = MakeEngine([], Date(2026, 3, 10, 10, 0));
        engine.RecordExternalFire(Guid.NewGuid(), Date(2026, 3, 10, 9, 0));
        Assert.Empty(engine.Events);
    }

    [Fact]
    public void ReconciledFireCountsTowardAdherence()
    {
        var created = Date(2026, 3, 10, 9, 0);
        var reminder = new Reminder
        {
            Title = "Water", Schedule = new Schedule.Interval(60), CreatedAt = created,
        };
        var (engine, _) = MakeEngine([reminder], Date(2026, 3, 10, 11, 0));

        engine.RecordExternalFire(reminder.Id, Date(2026, 3, 10, 10, 0));
        engine.Complete(reminder.Id);

        Assert.Equal(1.0, engine.Adherence(reminder.Id, created));
    }
}

File: Tests/ReminderCoreTests/ExternalFireTests.swift
import XCTest
@testable import ReminderCore

/// Tests for `recordExternalFire`, the reconciliation entry point used on iOS
/// where the system delivers pre-scheduled notifications and alarms while the
/// app is not running. It must produce the state a live `tick()` would have,
/// and it must be safe to call repeatedly with the same facts.
@MainActor
final class ExternalFireTests: XCTestCase {

    private var calendar: Calendar = {
        var cal = Calendar(identifier: .gregorian)
        cal.timeZone = TimeZone(identifier: "UTC")!
        return cal
    }()

    private func date(
        _ year: Int, _ month: Int, _ day: Int, _ hour: Int = 0, _ minute: Int = 0
    ) -> Date {
        var comps = DateComponents()
        comps.year = year; comps.month = month; comps.day = day
        comps.hour = hour; comps.minute = minute; comps.second = 0
        return calendar.date(from: comps)!
    }

    private func makeEngine(
        reminders: [Reminder],
        now: Date
    ) -> (ReminderEngine, MutableDateProvider) {
        let store = InMemoryDataStore(
            data: AppData(reminders: reminders, settings: Settings(), events: [])
        )
        let clock = MutableDateProvider(now: now)
        let engine = ReminderEngine(
            store: store, dateProvider: clock, presenter: nil, calendar: calendar
        )
        return (engine, clock)
    }

    func testStampsAnchorAndRecordsFiredEvent() {
        let created = date(2026, 3, 10, 9, 0)
        let reminder = Reminder(
            title: "Water", schedule: .interval(minutes: 60), createdAt: created
        )
        let (engine, _) = makeEngine(reminders: [reminder], now: date(2026, 3, 10, 11, 0))

        let stamp = date(2026, 3, 10, 10, 0)
        engine.recordExternalFire(id: reminder.id, at: stamp)

        XCTAssertEqual(engine.reminder(withID: reminder.id)?.lastFiredAt, stamp)
        XCTAssertEqual(engine.events.filter { $0.outcome == .fired }.count, 1)
        XCTAssertEqual(engine.events.first?.date, stamp)
    }

    func testRepeatedReconciliationDoesNotDuplicate() {
        let created = date(2026, 3, 10, 9, 0)
        let reminder = Reminder(
            title: "Water", schedule: .interval(minutes: 60), createdAt: created
        )
        let (engine, _) = makeEngine(reminders: [reminder], now: date(2026, 3, 10, 11, 0))

        let stamp = date(2026, 3, 10, 10, 0)
        engine.recordExternalFire(id: reminder.id, at: stamp)
        engine.recordExternalFire(id: reminder.id, at: stamp)
        engine.recordExternalFire(id: reminder.id, at: stamp)

        XCTAssertEqual(engine.events.filter { $0.outcome == .fired }.count, 1)
    }

    func testOlderStampThanExistingAnchorIsIgnored() {
        let created = date(2026, 3, 10, 9, 0)
        var reminder = Reminder(
            title: "Water", schedule: .interval(minutes: 60), createdAt: created
        )
        reminder.lastFiredAt = date(2026, 3, 10, 11, 0)
        let (engine, _) = makeEngine(reminders: [reminder], now: date(2026, 3, 10, 11, 30))

        engine.recordExternalFire(id: reminder.id, at: date(2026, 3, 10, 10, 0))

        XCTAssertEqual(
            engine.reminder(withID: reminder.id)?.lastFiredAt, date(2026, 3, 10, 11, 0),
            "An anchor must never move backwards"
        )
        XCTAssertTrue(engine.events.isEmpty)
    }

    func testElapsedSnoozeIsConsumedByTheFireThatHonouredIt() {
        let created = date(2026, 3, 10, 9, 0)
        var reminder = Reminder(
            title: "Tilt", schedule: .interval(minutes: 60), createdAt: created
        )
        reminder.snoozedUntil = date(2026, 3, 10, 9, 40)
        let (engine, _) = makeEngine(reminders: [reminder], now: date(2026, 3, 10, 10, 0))

        engine.recordExternalFire(id: reminder.id, at: date(2026, 3, 10, 9, 40))

        XCTAssertNil(engine.reminder(withID: reminder.id)?.snoozedUntil)
        XCTAssertEqual(
            engine.reminder(withID: reminder.id)?.lastFiredAt, date(2026, 3, 10, 9, 40)
        )
    }

    func testFutureSnoozeSurvivesAnEarlierExternalFire() {
        let created = date(2026, 3, 10, 9, 0)
        var reminder = Reminder(
            title: "Tilt", schedule: .interval(minutes: 60), createdAt: created
        )
        // The user snoozed *after* the notification was delivered; the snooze
        // is a promise about the future and must not be consumed by the past
        // fire being reconciled.
        reminder.snoozedUntil = date(2026, 3, 10, 12, 0)
        let (engine, _) = makeEngine(reminders: [reminder], now: date(2026, 3, 10, 10, 30))

        engine.recordExternalFire(id: reminder.id, at: date(2026, 3, 10, 10, 0))

        XCTAssertEqual(
            engine.reminder(withID: reminder.id)?.snoozedUntil, date(2026, 3, 10, 12, 0)
        )
    }

    func testUnknownReminderIsANoOp() {
        let (engine, _) = makeEngine(reminders: [], now: date(2026, 3, 10, 10, 0))
        engine.recordExternalFire(id: UUID(), at: date(2026, 3, 10, 9, 0))
        XCTAssertTrue(engine.events.isEmpty)
    }

    func testReconciledFireCountsTowardAdherence() {
        let created = date(2026, 3, 10, 9, 0)
        let reminder = Reminder(
            title: "Water", schedule: .interval(minutes: 60), createdAt: created
        )
        let (engine, _) = makeEngine(reminders: [reminder], now: date(2026, 3, 10, 11, 0))

        engine.recordExternalFire(id: reminder.id, at: date(2026, 3, 10, 10, 0))
        engine.complete(id: reminder.id)

        XCTAssertEqual(
            engine.adherence(for: reminder.id, since: created), 1.0
        )
    }
}
